python/arcticdb/diagnostics_engine/fragmentation_diagnosis.py

```diff
@@ -0,0 +1,197 @@
+        " data is."
+    )
+    display(Markdown(description))
+
+    plt.figure(figsize=(16, 9))
+    plt.hist(df["rows"], bins=100, rwidth=10)
+    plt.axvline(segment_row_size, color="r", linestyle="--")
+    plt.xlabel("Number of rows per Segment")
+    plt.ylabel("Frequency")
+    plt.title("Histogram of Number of rows per Data Segment")
+    plt.show()
+
+
+def fragmentation_diagnosis(lib, symbol, as_of=None):
+    """
+    Analyzes and displays the fragmentation level for a specified symbol from a library.
+
+    The function checks the provided library's type, reads the index of the symbol, and calculates
+    various statistics such as the total segment size, fragmentation, row-wise and column-wise fragmentation.
+    Based on these statistics, it generates a markdown report and a histogram plot showing the distribution
+    of data points per segment.
+
+    Parameters:
+    -----------
+    lib : object
+        The library object in which the symbol is located.
+    symbol : str
+        The symbol for which the fragmentation analysis is to be performed.
+    as_of : int, optional
+        The version of the symbol for which the analysis is to be performed.
+        If not specified, the latest version is used.
+
+    Returns:
+    --------
+    None
+    Outputs a markdown report and a histogram plot detailing the fragmentation analysis.
+
+    Example:
+    --------
+    fragmentation_diagnosis(lib, 'my_symbol', 3)
+    """
+
+    lib = check_and_adapt_library(lib)
+    if lib is None:
+        return
+
+    if not check_symbol_exists(lib, symbol, as_of):
+        return
+
+    segment_row_size = lib._lib_cfg.lib_desc.version.write_options.segment_row_size
+    column_group_size = lib._lib_cfg.lib_desc.version.write_options.column_group_size
+    dynamic_schema = lib._lib_cfg.lib_desc.version.write_options.dynamic_schema
+
+    # Calculate the actual amount of data stored across all the segments
+    df = lib.read_index(symbol, as_of=as_of)
+    df_with_segment_info = calculate_and_add_segment_occupancy_info(df)
+    sum_data_points, sum_rows, sum_columns = calculate_sum_segments_occupancy_info(df_with_segment_info)
+
+    # Calculate the maximum amount of data that can be stored across all segments
+    total_segment_size = segment_row_size * column_group_size
+    max_data_points, max_rows, max_columns = calculate_segments_occupancy_maximums(
+        df_with_segment_info, total_segment_size, segment_row_size, column_group_size
+    )
+
+    # Infer fragmentation
+    fragmentation, per_row_fragmentation, per_column_fragmentation = calculate_fragmentation(
+        sum_data_points, sum_rows, sum_columns, max_data_points, max_rows, max_columns, dynamic_schema
+    )
+
+    # Display Results
+    display_message(fragmentation, per_row_fragmentation, per_column_fragmentation)
+    plot_data_points(df, total_segment_size)
+    if not dynamic_schema:
+        plot_per_row_fragmentation(df, segment_row_size)
```

Comment:
This logic is wrong. For dynamic schema, it only makes sense to plot the row fragmentation, not the total fragmentation.
With static schema we should plot both.

Reply:
Idem. Done the changes.